device.go

```diff
@@ -229,37 +231,174 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
+func nvdimmDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.ContainerPath)
 }
 
-// updateSpecDeviceList takes a device description provided by the caller,
-// trying to find it on the guest. Once this device has been identified, the
-// "real" information that can be read from inside the VM is used to update
-// the same device in the list of devices provided through the OCI spec.
-// This is needed to update information about minor/major numbers that cannot
-// be predicted from the caller.
-func updateSpecDeviceList(device pb.Device, spec *pb.Spec) error {
-	// If no ContainerPath is provided, we won't be able to match and
-	// update the device in the OCI spec device list. This is an error.
-	if device.ContainerPath == "" {
+func vfioRebindDevice(bdf string) error {
+	fieldLogger := agentLog.WithField("guest-bdf", bdf)
+	fieldLogger.Debug("VFIO: Attempting to rebind")
+
+	sysPath := filepath.Join(sysBusPrefix, bdf)
+
+	// First we have to wait for the device to appear in the
+	// guest, since the hotplug can take a while.  We can't use
+	// the waiting logic in getDeviceName here, because the udev
+	// devices will only appear as we expect *after* we've
+	// rebound.
+	start := time.Now()
+	for {
+		_, err := os.Stat(sysPath)
+		if err == nil {
+			break
+		}
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("Error accessing %s: %s", sysPath, err)
+		}
+		if elapsed := time.Since(start); elapsed > vfioTimeout {
+			return fmt.Errorf("Timeout waiting for %s (%v)", sysPath, elapsed)
+		}
+		time.Sleep(vfioHotplugPollInterval)
+	}
+	fieldLogger.Debugf("VFIO: Found PCI device after %v", time.Since(start))
+
+	overridePath := filepath.Join(sysPath, "driver_override")
+	err := ioutil.WriteFile(overridePath, []byte("vfio-pci"), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", overridePath, err)
+	}
+
+	driverPath := filepath.Join(sysPath, "driver")
+	driver, err := os.Readlink(driverPath)
+	if os.IsNotExist(err) {
+		// This just means the device isn't bound to any driver
+		driver = ""
+	} else if err != nil {
+		return fmt.Errorf("Failed to readlink %s: %s", driverPath, err)
+	}
+
+	fieldLogger.Debugf("VFIO: Device previously bound to '%s'", driver)
+
+	if driver != "vfio-pci" && driver != "" {
+		fieldLogger.Debugf("VFIO: Unbinding from %s", driver)
+		unbindPath := filepath.Join(sysPath, bdf, "driver", "unbind")
+		err = ioutil.WriteFile(unbindPath, []byte(bdf), 0200)
+		if err != nil {
+			return fmt.Errorf("Failed to write %s: %s:", unbindPath, err)
+		}
+	}
+
+	err = ioutil.WriteFile(pciProbeFile, []byte(bdf), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", pciProbeFile, err)
+	}
+
+	fieldLogger.Info("VFIO: Rebind complete")
+	return nil
+}
+
+func vfioDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
```

Comment:
Please document this function a little better, in particular what it expects on input. I'm especially interested in the split of responsibilities between whoever generates the spec, the runtime and the agent, in order who knows what about the guest PCI devices, PCI bridges, IOMMU groups, etc.

Reply:
Um... the same as every other *Handler function...

I don't think it's reasonable to expect me to fix every underdocumented piece of the surrounding code before adding the new stuff.

Reply:
No, I don't think it's the same thing at all. Here, you are constructing a host to guest mapping that does not exist for the other cases. This difference is highlighted by the fact that the `sandbox` struct contains two maps dedicated to your specific algorithm, `vfioDeviceMap` and `sysToDevMap`, which are not needed (or only marginally) for other handlers.

As I indicated in my comment, I'm not really interested in what goes in what parameter or whatnot, but really more in the overall flow of data. The way I understand it, we have the following, which is so thoroughly unfamiliar to anybody who did not focus on the problem for one month that I am more likely than not to get it wrong:

- Some external entity like a high-level network management service assigns a VF to this container

- This is passed down in the manifest through an envrionment variable in the form `PCIDEVICE_foo=bar`, where `foo` and `bar` are something I already entirely forgot. But basically the workload know `foo`, and from there can get `bar` which I believe is generally speaking a `/dev/` entry.

- Since they are host BDFs, they are woefully inadequate for use in the guest. So here comes your magic.

- Underneath our VM is qemu. qemu is the one that knows what the host BDF will be mapped to in the guest. So we query qemu to get that information. Who does? I guess it's the runtime sending some qmp requests, though I have not fully reviewed that code yet.

- Well, at that stage, qemu has no idea about `/dev` entries in the guest. All it knows is hardware-level stuff, like PCI layout. So the best we get is a PCI B:D.F. The agent will have to deal with figuring out the `/dev` entry from that.

- The runtime then decorates the OCI spec with additional information about host to guest mappings. That comes in the form of device options that look like `B:D.F=guestBridge/guestDeviceID`. (Not sure about other options…)

- The agent then rebinds the device using the guestDeviceID to look it up under guest `/sys`

- The agent then waits for the device to show up as `/dev/vfio/NN` listening to udev event.

- When that happened, it rewrites the environment, specifically the `PCIDEVICE_foo=bar` variables so that they now correctly point to guest entities.

Is that a good summary? If so, could you convert that into a comment, with appropriate corrections, and with correct function names in the runtime and agent?

Reply:
> No, I don't think it's the same thing at all. Here, you are constructing a host to guest mapping that does not exist for the other cases. This difference is highlighted by the fact that the `sandbox` struct contains two maps dedicated to your specific algorithm, `vfioDeviceMap` and `sysToDevMap`, which are not needed (or only marginally) for other handlers.

Sorry, I was meaning specifically w.r.t. "what it expects on input".  It's true this is rather more complex than the other handlers.  `sysToDevMap` is absolutely used by the other handlers, to exactly the same degree it is here.

> As I indicated in my comment, I'm not really interested in what goes in what parameter or whatnot, but really more in the overall flow of data. The way I understand it, we have the following, which is so thoroughly unfamiliar to anybody who did not focus on the problem for one month that I am more likely than not to get it wrong:
> 
>     * Some external entity like a high-level network management service assigns a VF to this container
> 
>     * This is passed down in the manifest through an envrionment variable in the form `PCIDEVICE_foo=bar`, where `foo` and `bar` are something I already entirely forgot. But basically the workload know `foo`, and from there can get `bar` which I believe is generally speaking a `/dev/` entry.

In part.  The environment variable is set in order to inform the container payload of which device to use, it's not really intended to have meaning to the container runtime itself - but we need to thinking about it because of the address change.  For the benefit of the container runtime, the OCI spec includes both the VFIO control device `/dev/vfio/vfio` and one or more VFIO group devices `/dev/vfio/$NUMBER` in the linux/devices section of the config.json.

>     * Since they are host BDFs, they are woefully inadequate for use in the guest. So here comes your magic.
> 
>     * Underneath our VM is qemu. qemu is the one that knows what the host BDF will be mapped to in the guest. So we query qemu to get that information. Who does? I guess it's the runtime sending some qmp requests, though I have not fully reviewed that code yet. (Update: See proposed changes here: [kata-containers/runtime#2963](https://github.com/kata-containers/runtime/pull/2963))

Actually qemu doesn't know the guest BDF either, because it depends on the guest OS's (or firmware's) enumeration of the PCI bus.  PCI domains have their own cans of worms, though we can sidestep that for now since we're only considering PC-like machines which only use a single domain / host bridge.

What qemu *does* know is the PCI slot & function of the device, and each PCI bridge used to get to it.  We can use those to construct a "PCI path" which is meaningful to both host and guest.  The "bridgeSlot/deviceSlot" format is basically this - just abbreviated because we happen to know that we always place the devices under exactly one bridge (I do hope to generalize this at some point, but that's Kata2 work).

So, kata-runtime queries the slot for the device and its bridge from qemu, and what it sends to the agent is a mapping from host BDF to guest "PCI path".

>     * Well, at that stage, qemu has no idea about `/dev` entries in the guest. All it knows is hardware-level stuff, like PCI layout. So the best we get is a PCI B:D.F. The agent will have to deal with figuring out the `/dev` entry from that.

As above, not even a PCI B:D.F, only a PCI path.

>     * The runtime then decorates the OCI spec with additional information about host to guest mappings. That comes in the form of device options that look like `B:D.F=guestBridge/guestDeviceID`. (Not sure about other options…)

It's not really decorating the OCI spec, this is additional information passed in parallel with the OCI spec when we make the CreateContainer grpc call to the agent.

>     * The agent then rebinds the device using the guestDeviceID to look it up under guest `/sys`

First the agent has to resolve the PCI paths to guest side BDFs  Then it can rebind the device via `/sys`.

>     * The agent then waits for the device to show up as `/dev/vfio/NN` listening to udev event.

Yes.
>     * When that happened, it rewrites the environment, specifically the `PCIDEVICE_foo=bar` variables so that they now correctly point to guest entities.

Yes.

> Is that a good summary? If so, could you convert that into a comment, with appropriate corrections, and with correct function names in the runtime and agent?

Ok in outline, some errors in details.  I'll see what I can do about putting it into comment form.